client/state/plugins/installed/selectors.js

```diff
@@ -66,65 +66,131 @@ export function requestPluginsError( state ) {
 	return state.plugins.installed.requestError;
 }
 
-function getPluginsSelector( state, siteIds, pluginFilter ) {
-	let pluginList = reduce(
-		siteIds,
-		( memo, siteId ) => {
-			if ( isRequesting( state, siteId ) ) {
-				return memo;
-			}
+const getSiteIdsThatHavePlugins = createSelector(
+	( state ) => {
+		return Object.keys( state.plugins.installed.plugins ).map( ( siteId ) => Number( siteId ) );
+	},
+	( state ) => [ state.plugins.installed.plugins ]
+);
+
+/**
+ * The server returns plugins store at state.plugins.installed.plugins are indexed by site, which means
+ * that the information for a plugin may be spread across multiple site objects. This selector transforms
+ * that structure into one indexed by the plugin slugs and memoizes that structure.
+ */
+export const getAllPluginsIndexedByPluginSlug = createSelector(
+	( state ) => {
+		if ( isRequestingForAllSites( state ) ) {
+			return {};
+		}
 
-			// We currently support fetching plugins per site and also fetching all plugins
-			// in bulk, aiming to optimize the UX in some flows.
-			if ( isRequestingForAllSites( state ) ) {
-				return memo;
+		return getSiteIdsThatHavePlugins( state ).reduce( ( plugins, siteId ) => {
+			if ( isRequesting( state, siteId ) ) {
+				return plugins;
 			}
 
-			const list = state.plugins.installed.plugins[ siteId ] || [];
-			list.forEach( ( item ) => {
-				const sitePluginInfo = pick( item, [ 'active', 'autoupdate', 'update', 'version' ] );
+			const pluginsForSite = state.plugins.installed.plugins[ siteId ] || [];
+			pluginsForSite.forEach( ( plugin ) => {
+				const sitePluginInfo = pick( plugin, [ 'active', 'autoupdate', 'update', 'version' ] );
 
-				memo[ item.slug ] = {
-					...memo[ item.slug ],
-					...item,
+				plugins[ plugin.slug ] = {
+					...plugins[ plugin.slug ],
+					...plugin,
 					sites: {
-						...memo[ item.slug ]?.sites,
+						...plugins[ plugin.slug ]?.sites,
 						[ siteId ]: sitePluginInfo,
 					},
 				};
 			} );
-			return memo;
-		},
-		{}
-	);
+			return plugins;
+		}, {} );
+	},
+	( state ) => [
+		isRequestingForAllSites( state ),
+		getSiteIdsThatHavePlugins( state ),
+		...getSiteIdsThatHavePlugins( state ).map( ( siteId ) => isRequesting( state, siteId ) ),
+	]
+);
 
-	if ( pluginFilter && _filters[ pluginFilter ] ) {
-		pluginList = filter( pluginList, _filters[ pluginFilter ] );
-	}
+/**
+ * The plugins here differ from the plugin objects found on state.plugins.installed.plugins in that each plugin
+ * object has data from all the different sites on it, whereas on state.plugins.installed.plugins they only have
+ * the state for the site which index they are under. The objects here are the same as those returned from
+ * getAllPluginsIndexedByPluginSlug, except they are indexed by siteId.
+ */
+export const getAllPluginsIndexedBySiteId = createSelector(
```

Comment:
This selector restructures the data calculated in getAllPluginsIndexedByPluginSlug for O(1) if you have a site ID, and memoizes that.